src/main/java/com/minecolonies/coremod/util/InventoryUtils.java

```diff
@@ -294,603 +363,1316 @@ public static boolean hasToolLevel(final String tool, final ItemStack stack, fin
+     * Returns the index of the first occurrence of an ItemStack that matches
+     * the given predicate in the {@link ICapabilityProvider}.
      *
-     * @param inventory Inventory to clear.
+     * @param provider                    Provider to check
+     * @param itemStackSelectionPredicate The predicate to match.
+     * @return Index of the first occurrence
      */
-    public static void clear(@NotNull final IInventory inventory)
+    public static int findFirstSlotInProviderNotEmptyWith(@NotNull final ICapabilityProvider provider, Predicate<ItemStack> itemStackSelectionPredicate)
     {
-        for (int slot = 0; slot < inventory.getSizeInventory(); slot++)
+        for (IItemHandler handler : getItemHandlersFromProvider(provider))
         {
-            inventory.setInventorySlotContents(slot, ItemStack.EMPTY);
+            int foundSlot = findFirstSlotInItemHandlerNotEmptyWith(handler, itemStackSelectionPredicate);
+            if (foundSlot > -1)
+            {
+                return foundSlot;
+            }
         }
+
+        return -1;
     }
 
     /**
-     * Returns a slot number if an inventory contains given tool type.
+     * Returns the index of the first occurrence of an ItemStack that matches
+     * the given predicate in the {@link IItemHandler}.
+     * Also applies the not empty check.
      *
-     * @param inventory the inventory to get the slot from.
-     * @param tool      the tool type to look for.
-     * @return slot number if found, -1 if not found.
+     * @param itemHandler                 ItemHandler to check
+     * @param itemStackSelectionPredicate The predicate to match.
+     * @return Index of the first occurrence
      */
-    public static int getFirstSlotContainingTool(@NotNull final IInventory inventory, @NotNull final String tool)
+    public static int findFirstSlotInItemHandlerNotEmptyWith(@NotNull final IItemHandler itemHandler, @NotNull final Predicate<ItemStack> itemStackSelectionPredicate)
     {
-        for (int i = 0; i < inventory.getSizeInventory(); i++)
+        @NotNull final Predicate<ItemStack> notEmptyPredicate = itemStack -> !InventoryUtils.isItemStackEmpty(itemStack);
+        notEmptyPredicate.and(itemStackSelectionPredicate);
+
+        for (int slot = 0; slot < itemHandler.getSlots(); slot++)
         {
-            final ItemStack item = inventory.getStackInSlot(i);
-            //Only classic fishingRod recognized as a fishingTool
-            if (item != null && (item.getItem().getToolClasses(item).contains(tool) || ("hoe".equals(tool) && item.getUnlocalizedName().contains("hoe"))
-                                   || ("rod".equals(tool) && item.getUnlocalizedName().contains("fishingRod"))))
+            if (notEmptyPredicate.test(itemHandler.getStackInSlot(slot)))
             {
-                return i;
+                return slot;
             }
         }
+
         return -1;
+        //TODO: Later harden contract to remove compare on slot := -1
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)